gprofiler/profilers/factory.py

```diff
@@ -24,44 +28,67 @@ def get_profilers(
+            if arch not in config.get_supported_archs() and len(requested_configs) == 1:
                 logger.warning(f"Disabling {profiler_name} because it doesn't support this architecture ({arch})")
                 continue
-
-            if profiling_mode not in profiler_config.supported_profiling_modes:
+            if profiling_mode not in config.supported_profiling_modes:
                 logger.warning(
                     f"Disabling {profiler_name} because it doesn't support profiling mode {profiling_mode!r}"
                 )
                 continue
-
+            selected_configs.append(config)
+        if not selected_configs:
+            logger.warning(f"Disabling {runtime} profiling because no profilers were selected")
+            continue
+        # create instances of selected profilers one by one, select first that is ready
+        ready_profiler = None
+        runtime_arg_names = [arg.dest for config in get_profilers_registry()[runtime] for arg in config.profiler_args]
+        for profiler_config in selected_configs:
+            profiler_name = profiler_config.profiler_name
             profiler_kwargs = profiler_init_kwargs.copy()
+            profiler_arg_names = [arg.dest for arg in get_profiler_arguments(runtime, profiler_name)]
             for key, value in user_args.items():
-                if key.startswith(lower_profiler_name) or key in COMMON_PROFILER_ARGUMENT_NAMES:
+                if (
+                    key in profiler_arg_names
+                    or key in COMMON_PROFILER_ARGUMENT_NAMES
+                    or key.startswith(runtime_args_prefix)
+                    and key not in runtime_arg_names
+                ):
                     profiler_kwargs[key] = value
             try:
                 profiler_instance = profiler_config.profiler_class(**profiler_kwargs)
+                if profiler_instance.check_readiness():
+                    ready_profiler = profiler_instance
+                    break
             except Exception:
-                logger.critical(
-                    f"Couldn't create the {profiler_name} profiler, not continuing."
-                    f" Run with --no-{profiler_name.lower()} to disable this profiler",
-                    exc_info=True,
-                )
-                sys.exit(1)
-            else:
-                if isinstance(profiler_instance, SystemProfiler):
-                    system_profiler = profiler_instance
-                else:
-                    process_profilers_instances.append(profiler_instance)
-
+                if len(requested_configs) == 1:
+                    logger.critical(
+                        f"Couldn't create the {profiler_name} profiler for runtime {runtime}, not continuing."
+                        f" Request different profiler for runtime with --{runtime_args_prefix}-mode, or disable"
+                        f" {runtime} profiling with --no-{runtime_args_prefix} to disable this profiler",
```

Comment:
```suggestion
                        f" {runtime} profiling with --{runtime_args_prefix}-mode=none to disable this profiler",
```

I prefer this way of expressing it (`--no-...` is legacy to me)